### PR TITLE
[LLVM] Add option to store Parent-pointer in ilist_node_base

### PR DESCRIPTION
**Pull request by SLTozer**
This patch adds a new option for `ilist`, `ilist_parent<ParentTy>`, that enables storing an additional pointer in the `ilist_node_base` type to a specified "parent" type, and uses that option for `Instruction`.

This is distinct from the `ilist_node_with_parent` class, despite their apparent similarities. The `ilist_node_with_parent` class is a subclass of `ilist_node` that requires its own subclasses to define a `getParent` method, which is then used by the owning `ilist` for some of its operations; it is purely an interface declaration. The `ilist_parent` option on the other hand is concerned with data; it is declared as a property of the list itself, and causes every node of that list to own a pointer to the list owner; it has no effect on node access and adds no new methods. Not all `ilist_node_with_parent`s should be in lists with `ilist_parent`; all lists with `ilist_parent` currently contain `ilist_node_with_parent`s, but there's no real advantage to enforcing this relationship at compile time.

Currently, we can use `BasicBlock::iterator` to insert instructions, _except_ when either the iterator is invalid (`NodePtr=0x0`), or when the iterator points to a sentinel value (`BasicBlock::end()`). The former can be easily verified from the iterator object itself, but the latter is currently not detectable (as we do not enable explicit sentinel tracking), and even if we know it we have no clean way to get the correct `BasicBlock*` value from the sentinel node (technically it should exist at a fixed offset from the start of the owning `BasicBlock` object, but using that fact is tricky without reinterpret_cast and funky pointer arithmetic). Moving the `BasicBlock *Parent` field from `Instruction` to `ilist_node_base` means that we can fetch the owning BB from the `end()` iterator, which allows us to use iterators for all insertions, without needing to store or pass an extra `BasicBlock *BB` argument alongside it.

The motivation for this is to move smoothly towards the [proposed rewrite of the instruction insertion interface](https://discourse.llvm.org/t/rfc-rewriting-the-instruction-insertion-interface/78494/1), which itself is needed to reduce the potential for errors to sneak in to the compiler from the use of insertion patterns that behave incorrectly with respect to debug records. This also allows us to greatly deduplicate the code for creating new instructions; in a followup commit (https://github.com/llvm/llvm-project/pull/94226), this behaviour is used to define a simple `InsertPosition` class that replaces the "insertBefore/insertAtEnd" arguments for all instruction creation/constructor methods, resulting in a significant reduction in the size of that file with [almost neutral performance impact](http://llvm-compile-time-tracker.com/compare.php?from=b2bd024384b484647da9fd9863bf6f77b5731949&to=3e5b1785fafacea2d3302d4ca9a4f85cb4cd808c&stat=instructions%3Au).

The performance impact of this patch can be seen on the compile time tracker [here](http://llvm-compile-time-tracker.com/compare.php?from=b2bd024384b484647da9fd9863bf6f77b5731949&to=f50c98bee78cec619e8e79f6386b70af5994abe2&stat=instructions:u). The results are basically neutral, except for a very slight increase in compile time (0.04%) and file size (0.01%) of stage 2 clang; the followup commit that exploits this change causes a larger reduction in both of these statistics.

### CODE REVIEW

File: llvm/include/llvm/ADT/ilist_node_base.h [hunk 1]
```diff
@@ -42,6 +45,48 @@ template <> class ilist_node_base<true> {
...
 
+  void setNodeBaseParent(void) {}
+  inline void getNodeBaseParent() const {}
+
+  bool isSentinel() const { return PrevAndSentinel.getInt(); }
+  bool isKnownSentinel() const { return isSentinel(); }
+  void initializeSentinel() { PrevAndSentinel.setInt(true); }
+};
+
+template <class ParentPtrTy> class ilist_node_base<false, ParentPtrTy> {
```
Comment (dexonsmith) [inline llvm/include/llvm/ADT/ilist_node_base.h:56]: Can this inherit from `<false, void>` to reduce code duplication? Or if that's confusing, extract a common base class for the previous/next logic?

**Review by dexonsmith — COMMENTED** [review]
I don't have time for a detailed review (travelling for the next few months)... but the plan in principle SGTM! Seems like a great way to detect the end iterator. Thanks for working on this.

File: llvm/include/llvm/ADT/ilist_node_base.h (hunk 1 above)
Comment (SLTozer) [reply llvm/include/llvm/ADT/ilist_node_base.h:56]: Unfortunately not directly, but the common base class approach could work!

**Review by nikic — COMMENTED** [review]
Not really familiar with all the infrastructure here, but conceptually this looks good to me. I've always disliked how a proper insertion point requires both an iterator and a BasicBlock, and this solves the problem quite elegantly.

**Comment by SLTozer** [comment]
Alright, I've made some modifications - using some mixin classes to 1) reduce a bit of duplication in `ilist_node_base`, and 2) make the parent access a bit "smarter": `ilist_node_impl` will now have public get/set parent methods iff the list uses `ilist_parent`, and similarly the `ilist_iterator` only provides `getNodeParent()` if the node type uses `ilist_parent`. This reduces the potential for misuse (unknowingly trying to `getNodeParent()` from an iterator that doesn't actually support it), and means that `Instruction` doesn't need to declare get/set parent at all. 

I think this is getting close to "too much template logic"; I also experimented making changes to `SymbolTableList` that would set the parent on the sentinel node automatically (as it does for other nodes), but it ended up at 30 lines of garbled template nonsense in order to avoid one call to `end().getNodePtr()->setParent(this)` in the `BasicBlock` constructor, so I decided not to add that to the patch.

**Comment by SLTozer** [comment]
Ping

**Comment by nikic** [comment]
@SLTozer There is a compilation failure in LLVM unit tests (IListBaseTest.cpp).

**Comment by SLTozer** [comment]
> @SLTozer There is a compilation failure in LLVM unit tests (IListBaseTest.cpp).

Ah, I missed the unit tests! Fixed the compilation errors and added some tests for the new ilist_node_base behaviour; I'll also add some unit tests for the other IList classes that got functional changes in this patch.

**Review by nikic — APPROVED** [review]
LGTM assuming this has no undue impact on build times.

**Comment by SLTozer** [comment]
> LGTM assuming this has no undue impact on build times.

The first test I ran reduced had no significant impact on compile times [(results)](http://llvm-compile-time-tracker.com/compare.php?from=b2bd024384b484647da9fd9863bf6f77b5731949&to=f50c98bee78cec619e8e79f6386b70af5994abe2&stat=instructions:u), other than a slight increase in the time to build clang; this increase in time-to-build-clang is more than compensated for by the followup patch for InsertPosition [(results)](http://llvm-compile-time-tracker.com/compare.php?from=f50c98bee78cec619e8e79f6386b70af5994abe2&to=3e5b1785fafacea2d3302d4ca9a4f85cb4cd808c&stat=instructions:u). I'll submit a quick check to the tracker to verify that the recent commits and rebase hasn't affected compile times, but most likely it will be fine. Thank you for the review!

**Comment by SLTozer** [comment]
Latest update, there's no major impact on compile times [(results)](http://llvm-compile-time-tracker.com/compare.php?from=9e63632b3274dc1b20502b569e79a311977e0a97&to=8095aff5f9a842994a36d085158c9d1e476cd5a9&stat=instructions:u); the only cost is the time to build clang, which as noted will be fixed by the next patch along.

**Comment by nikic** [comment]
> Latest update, there's no major impact on compile times [(results)](http://llvm-compile-time-tracker.com/compare.php?from=9e63632b3274dc1b20502b569e79a311977e0a97&to=8095aff5f9a842994a36d085158c9d1e476cd5a9&stat=instructions:u); the only cost is the time to build clang, which as noted will be fixed by the next patch along.

Thanks, those numbers look good to me.

**Comment by dklimkin** [comment]
H Stephen,

We ran into some issues with `getParent()` and `getNodeBaseParent()` returning a const pointer rather than a pointer to a const, resulting in iterators returned also not being const. I'll send a PR with a workaround but adding a note here in case you have a better solution in mind.

**Comment by SLTozer** [comment]
Ouch, that's an oversight on my part - I think the correct solution probably involves adding a `const_parent_ptr_ty` to `node_options`, though I think that either requires us to pass both around (making the template specialization logic a bit messier/nastier), or to pass around the parent type directly, which is probably the better choice - the only cost is that we can't set a `void*` parent pointer, but we probably don't need that!

**Comment by dklimkin** [comment]
Sent you https://github.com/llvm/llvm-project/pull/96053 but I'd prefer passing around the parent type directly. Adding to the options makes things messy, I did try that. Thanks!

**Pull request merged**